orm/src/main/java/org/hibernate/search/query/hibernate/impl/CriteriaObjectInitializer.java

```diff
@@ -105,33 +111,124 @@ private void setCriteriaTimeout(Criteria criteria, TimeoutManager timeoutManager
 		}
 	}
 
-	private void buildUpCriteria(EntityInfo[] entityInfos,
-			Criteria criteria,
-			int maxResults,
-			ObjectInitializationContext objectInitializationContext) {
+	/**
+	 * Returns a list with one or more {@link Criteria} objects for loading the given entity infos. The returned list
+	 * will contain one criteria object for each id space used by the given infos. A single criteria will be returned in
+	 * case all the entity infos originate from the same id space.
+	 */
+	private List<Criteria> buildUpCriteria(EntityInfo[] entityInfos, ObjectInitializationContext objectInitializationContext) {
+		Map<Class<?>, List<EntityInfo>> infosByIdSpace = groupInfosByIdSpace( entityInfos, objectInitializationContext );
+
+		// all entities from same id space -> single criteria
+		if ( infosByIdSpace.size() == 1 ) {
+			// no explicitly user specified criteria query, define one
+			Criteria criteria = objectInitializationContext.getCriteria();
+
+			if ( criteria == null ) {
+				criteria = objectInitializationContext.getSession().createCriteria( infosByIdSpace.keySet().iterator().next() );
+			}
+
+			criteria.add( getIdListCriterion( infosByIdSpace.values().iterator().next(), objectInitializationContext ) );
+
+			return Collections.singletonList( criteria );
+		}
+		// entities originate from different id spaces -> criteria per space
+		else {
+			// Cannot use external criteria for fetching entities from different spaces
+			if ( objectInitializationContext.getCriteria() != null ) {
+				log.givenCriteriaObjectCannotBeApplied();
```

Comment:
AFAICS we cannot make use of an externally provided `Criteria` in case we need to fetch entities from several id spaces (i.e. I don't think we can fetch all entities using a single criteria object in this case).

Unless there actually is a way to do this, I could imagine to provide a way for registering a call back which returns user-provided criteria for given entity types:

    public interface CriteriaProvider {
        Criteria getCriteria(EntityType entityType);
    }

and then:

    final FullTextSession fullTextSession = ...;
    List results = fullTextSession.createFullTextQuery( query )
        .setCriteriaProvider(
            new CriteriaProvider {
                public Criteria getCriteria(EntityType entityType) {
                    // for Book add a fetch join, everything else is defaulted
                    if ( entityType == Book.class ) {
                        return fullTextSession.createCriteria( Book.class ).setFetchMode( "authors", FetchMode.JOIN );
                    }
                    else {
                        return fullTextSession.createCriteria( entityType );
                    }
                }
            }
        )
        .list();

This callback would be invoked by the initializer if requiring a `Criteria` for a given type.

Any thoughts? Anything much simpler I didn't see?